pkg/utils/apply/apply.go

```diff
@@ -0,0 +1,205 @@
+// nolint: golint
+type ApplyOption func(ctx context.Context, existing, desired runtime.Object) error
+
+// NewAPIApplicator creates returns an Applicator that applies changes to an
+// object by either creating or updating/patching it in a Kubernetes API server.
+func NewAPIApplicator(c client.Client) *APIApplicator {
+	return &APIApplicator{
+		creatorFn(createOrGetExisting),
+		patcherFn(threeWayMergePatch),
+		c}
+}
+
+type creator interface {
+	createOrGetExisting(context.Context, client.Client, runtime.Object, ...ApplyOption) (runtime.Object, error)
+}
+
+type creatorFn func(context.Context, client.Client, runtime.Object, ...ApplyOption) (runtime.Object, error)
+
+func (fn creatorFn) createOrGetExisting(ctx context.Context, c client.Client, o runtime.Object, ao ...ApplyOption) (runtime.Object, error) {
+	return fn(ctx, c, o, ao...)
+}
+
+type patcher interface {
+	calculatePatch(c, m runtime.Object) (client.Patch, error)
+}
+
+type patcherFn func(c, m runtime.Object) (client.Patch, error)
+
+func (fn patcherFn) calculatePatch(c, m runtime.Object) (client.Patch, error) {
+	return fn(c, m)
+}
+
+// APIApplicator implements Applicator
+type APIApplicator struct {
+	creator
+	patcher
+	c client.Client
+}
+
+// UpdatingApply applies changes to an object with updating
+func (a *APIApplicator) UpdatingApply(ctx context.Context, desired runtime.Object, ao ...ApplyOption) error {
+	existing, err := a.createOrGetExisting(ctx, a.c, desired, ao...)
+	if err != nil {
+		return err
+	}
+	if existing == nil {
+		return nil
+	}
+
+	// the object already exists, update
+	if err := applyApplyOptions(ctx, existing, desired, ao); err != nil {
+		return err
+	}
+	desired.(Object).SetResourceVersion(existing.(Object).GetResourceVersion())
+	return errors.Wrap(a.c.Update(ctx, desired), "cannot update object")
+}
+
+// PatchingApply applies changes to an object with patching
+func (a *APIApplicator) PatchingApply(ctx context.Context, desired runtime.Object, ao ...ApplyOption) error {
+	existing, err := a.createOrGetExisting(ctx, a.c, desired, ao...)
+	if err != nil {
+		return err
+	}
+	if existing == nil {
+		return nil
+	}
+
+	// the object already exists, patch changes
+	if err := applyApplyOptions(ctx, existing, desired, ao); err != nil {
+		return err
+	}
+	patch, err := a.patcher.calculatePatch(existing, desired)
+	if err != nil {
+		return errors.Wrap(err, "cannot calculate patch by computing a three way diff")
+	}
+	return errors.Wrapf(a.c.Patch(ctx, desired, patch), "cannot patch object")
+}
+
+// createOrGetExisting will create the object if it does not exist
+// or return the existing object if it already exists
+func createOrGetExisting(ctx context.Context, c client.Client, desired runtime.Object, ao ...ApplyOption) (runtime.Object, error) {
+	m, ok := desired.(Object)
+	if !ok {
+		return nil, errors.New("cannot access object metadata")
+	}
+
+	if m.GetName() == "" && m.GetGenerateName() == "" {
```

Comment:
can we create the object successfully if this case exist? Maybe k8s apiserver will complain like `resource without a name can't be created`

Reply:
it's used for creating objects with only generateName. I made a mistake here. it should be 
`m.GetGenerateName() != "" ` I will fix it.